File: backend/app/schemas/stock.py
"""
Pydantic 请求/响应模型
"""
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel


# ====== 股票搜索 ======

class StockBrief(BaseModel):
    code: str
    name: str
    market: str
    industry: Optional[str] = None

    class Config:
        from_attributes = True


class StockSearchResponse(BaseModel):
    query: str
    results: list[StockBrief]
    total: int = 0


# ====== 股票详情 ======

class FinancialItem(BaseModel):
    report_period: str
    revenue: Optional[float] = None
    net_profit: Optional[float] = None
    total_assets: Optional[float] = None
    total_equity: Optional[float] = None
    operating_cash_flow: Optional[float] = None
    receivables: Optional[float] = None
    inventory: Optional[float] = None
    eps: Optional[float] = None
    pe_ratio: Optional[float] = None
    pb_ratio: Optional[float] = None

    # 计算字段
    gross_margin: Optional[float] = None
    net_margin: Optional[float] = None
    roe: Optional[float] = None
    debt_ratio: Optional[float] = None
    revenue_growth_yoy: Optional[float] = None
    profit_growth_yoy: Optional[float] = None


class AnnouncementItem(BaseModel):
    id: int
    title: str
    publish_date: date
    summary: Optional[str] = None
    url: Optional[str] = None


class QuickAnalysisResult(BaseModel):
    summary: str
    strengths: list[str] = []
    risks: list[str] = []
    metrics_commentary: str = ""


class StockDetailResponse(BaseModel):
    code: str
    name: str
    market: str
    industry: Optional[str] = None
    listing_date: Optional[date] = None
    is_active: bool = True
    financials: list[FinancialItem] = []
    announcements: list[AnnouncementItem] = []
    quick_analysis: Optional[QuickAnalysisResult] = None


# ====== AI 分析 ======

class QuickAnalysisResponse(BaseModel):
    code: str
    analysis: dict  # 灵活接受 AI 返回的结构


class DeepAnalysisRequest(BaseModel):
    focus: str = "full"  # full / financial / risk


class DeepAnalysisResponse(BaseModel):
    code: str
    task_id: str


# ====== 报告 ======

class ReportGenerateRequest(BaseModel):
    template: str = "standard"


class ReportGenerateResponse(BaseModel):
    code: str
    task_id: str


class ReportStatusResponse(BaseModel):
    task_id: str
    status: str  # processing / completed / failed
    title: Optional[str] = None
    progress: int = 0
    message: str = ""


class ReportResponse(BaseModel):
    report_id: int
    title: str
    content: str
    report_type: str
    status: str
    created_at: datetime


class ReportListItem(BaseModel):
    report_id: int
    title: str
    report_type: str
    status: str
    created_at: datetime


# ====== 自选股 ======

class WatchlistItemResponse(BaseModel):
    id: int
    code: str
    name: str
    market: str
    industry: Optional[str] = None
    added_at: datetime
    quick_summary: Optional[str] = None
    has_alert: bool = False


class WatchlistAddRequest(BaseModel):
    code: str


class WatchlistRemoveResponse(BaseModel):
    success: bool
    message: str

File: frontend/src/types/stock.ts
/**
 * 股票相关类型定义
 */

export interface StockBrief {
  code: string
  name: string
  market: 'SZ' | 'SH' | 'BJ'
  industry?: string
}

export interface FinancialItem {
  report_period: string
  revenue: number | null
  net_profit: number | null
  total_assets: number | null
  total_equity: number | null
  operating_cash_flow: number | null
  receivables: number | null
  inventory: number | null
  eps: number | null
  pe_ratio: number | null
  pb_ratio: number | null
  gross_margin: number | null
  net_margin: number | null
  roe: number | null
  debt_ratio: number | null
  revenue_growth_yoy: number | null
  profit_growth_yoy: number | null
}

export interface AnnouncementItem {
  id: number
  title: string
  publish_date: string
  summary: string | null
  url: string | null
}

export interface QuickAnalysisResult {
  summary: string
  strengths: string[]
  risks: string[]
  metrics_commentary: string
}

export interface StockDetail {
  code: string
  name: string
  market: string
  industry: string | null
  listing_date: string | null
  is_active: boolean
  financials: FinancialItem[]
  announcements: AnnouncementItem[]
  quick_analysis: QuickAnalysisResult | null
}

export interface ReportInfo {
  report_id: number
  title: string
  content: string
  report_type: string
  status: string
  created_at: string
}

export interface WatchlistItem {
  id: number
  code: string
  name: string
  market: string
  industry: string | null
  added_at: string
  quick_summary: string | null
  has_alert: boolean
}

export interface SearchResponse {
  query: string
  results: StockBrief[]
}

// ====== 异动监控 ======

export interface AlertItem {
  id: number
  stock_code: string
  stock_name: string
  alert_type: 'announcement' | 'shareholder' | 'regulatory' | 'financial'
  severity: 'high' | 'medium' | 'low'
  title: string
  description: string
  is_read: boolean
  detected_at: string
}

// ====== 行业对标 ======

export interface IndustryStats {
  avg: number
  median: number
  max: number
  min: number
  count: number
}

export interface IndustryRanking {
  rank: number
  total: number
  percentile: number
}

export interface IndustryPeer {
  code: string
  name: string
  report_period?: string
  revenue?: number | null
  net_profit?: number | null
  net_margin?: number | null
  roe?: number | null
  debt_ratio?: number | null
  revenue_growth_yoy?: number | null
  profit_growth_yoy?: number | null
}

export interface IndustryComparison {
  code: string
  name: string
  industry: string
  target_metrics: {
    report_period: string
    revenue: number | null
    net_profit: number | null
    net_margin: number | null
    roe: number | null
    debt_ratio: number | null
    revenue_growth_yoy: number | null
    profit_growth_yoy: number | null
  }
  industry_stats: Record<string, IndustryStats>
  rankings: Record<string, IndustryRanking>
  peer_count: number
  top_peers: IndustryPeer[]
  error?: string
}


// ====== 行情 / 技术面 / 资金面 ======

export interface MarketSnapshot {
  code: string
  latest: {
    date: string | null
    close: number | null
    pct_chg: number | null
    turnover: number | null
    chg_5d: number | null
  }
  indicators: {
    ma5?: number | null
    ma10?: number | null
    ma20?: number | null
    ma60?: number | null
    macd_dif?: number | null
    macd_dea?: number | null
    macd?: number | null
    macd_cross?: 'golden' | 'dead' | 'none'
    kdj_k?: number | null
    kdj_d?: number | null
    kdj_j?: number | null
    rsi?: number | null
    series?: { date: string; close: number; dif: number; dea: number; macd: number }[]
  }
  fund_flow: {
    signal?: 'accumulate' | 'distribute' | 'neutral'
    label?: string
    main_net_5d?: number
    main_net_today?: number
    main_net_pct_today?: number | null
    positive_days_5d?: number
    series?: { date: string; main_net: number; pct_chg: number | null }[]
  }
  lhb: {
    date: string
    reason: string
    interpret: string
    net_buy: number | null
    pct_chg: number | null
    after_1d: number | null
    after_5d: number | null
  }[]
}

export interface ShortTermResult {
  rating: string
  summary: string
  technical: string
  capital: string
  opportunities: string[]
  risks: string[]
  lhb_note: string
}
